src/v/datalake/schema_avro.cc

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#include "datalake/schema_avro.h"
+
+#include <seastar/util/defer.hh>
+
+#include <avro/Schema.hh>
+namespace datalake {
+
+namespace {
+static constexpr size_t max_schema_depth = 100;
+struct state {
+    std::deque<avro::NodePtr> type_hierarchy;
+    iceberg::nested_field::id_t last_field_id{0};
+
+    iceberg::nested_field::id_t next_field_id() { return last_field_id++; }
+};
+
+conversion_outcome<std::optional<iceberg::field_type>>
+field_type_from_avro(const avro::NodePtr& node, state& state);
+
+conversion_outcome<iceberg::struct_type>
+struct_from_avro_record(const avro::NodePtr& node, state& state) {
+    iceberg::struct_type ret;
+    ret.fields.reserve(node->leaves());
+    for (size_t i = 0; i < node->leaves(); ++i) {
+        auto leaf = node->leafAt(i);
+
+        auto result = field_type_from_avro(leaf, state);
+        if (result.has_error()) {
+            return result.error();
+        }
+
+        auto field = std::move(result.value());
+        if (field.has_value()) {
+            ret.fields.push_back(iceberg::nested_field::create(
+              state.next_field_id(),
+              node->nameAt(i),
+              iceberg::field_required::yes,
+              std::move(field.value())));
+        }
+    }
+    return ret;
+}
+
+iceberg::decimal_type create_decimal(const avro::NodePtr& node) {
+    return iceberg::decimal_type{
+      .precision = static_cast<uint32_t>(node->logicalType().precision()),
+      .scale = static_cast<uint32_t>(node->logicalType().scale())};
+}
+
+bool is_recursive_type(const state& state, const avro::NodePtr& node) {
+    for (auto& n : state.type_hierarchy) {
+        if (node == n) {
```

Comment:
is it unnecessary to descend into the type hierarchy vs looking at the top-level?